### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by mstaring**
ENH: merging work of yqiao

This contains the PhD research by Yuchuan Qiao.
Part of this commit came from mstaring/performance_ITK4 and part from YuchuanQiao/performance_ITK4
I took it manually and committed it.

### CODE REVIEW

**Comment by mstaring** [comment]
@N-Dekker @YuchuanQiao This commit contains work of Yuchuan from his PhD thesis. Could you please check i) if this code is itk5 compliant, ii) if I copied it correctly from https://github.com/YuchuanQiao/elastix/tree/performance_ITK4, because I did that manually.

**Comment by N-Dekker** [comment]
@mstaring The "azure" CI has accepted the work of @YuchuanQiao for all tested platforms, including ITK5. Is it included (for compilation and test) "by default", or it there a CMake flag to "enable" his work?

File: Common/itkComputePreconditionerUsingDisplacementDistribution.h [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+
+
+namespace itk
+{
+/**\class ComputePreconditionerUsingDisplacementDistribution
+ * \brief This is a helper class for the automatic estimation of a preconditioner for the FPSGD optimizer.
+ * // update below
+ * More specifically this class computes the Jacobian terms related to the automatic
+ * parameter estimation for the adaptive stochastic gradient descent optimizer.
+ * Details can be found in the SPIE paper
```
Comment (YuchuanQiao) [inline Common/itkComputePreconditionerUsingDisplacementDistribution.h]: TMI paper

File: Common/itkComputePreconditionerUsingDisplacementDistribution.h [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+ *
+ * [1]  Y.Qiao, B.P.F. Lelieveldt, M.Staring
```
Comment (YuchuanQiao) [inline Common/itkComputePreconditionerUsingDisplacementDistribution.h]: Qiao Y, Lelieveldt BP, Staring M. An efficient preconditioner for stochastic gradient descent optimization of image registration. IEEE transactions on medical imaging. 2019 Feb 11.

File: Components/Optimizers/PreconditionedGradientDescent/itkAdaptiveStochasticPreconditionedGradientDescentOptimizer.h [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+ *
+ *     \f[ t_{k+1} = [ t_k + sigmoid( -g_k^T P g_{k-1} ) ]^+ \f]
+ *
+ * where \f$g_k\f$ equals \f$dC/dx\f$ at iteration \f$k\f$.
+ * For \f$t_0\f$ the InitialTime is used, which is defined in the
+ * the superclass (StandardGradientDescentOptimizer). Whereas in the
+ * superclass this parameter is superfluous, in this class it makes sense.
+ *
+ * This method is described in the following references:
+ *
```
Comment (YuchuanQiao) [inline Components/Optimizers/PreconditionedGradientDescent/itkAdaptiveStochasticPreconditionedGradientDescentOptimizer.h]: Qiao Y, Lelieveldt BP, Staring M. An efficient preconditioner for stochastic gradient descent optimization of image registration. IEEE transactions on medical imaging. 2019 Feb 11.

File: Components/Optimizers/PreconditionedGradientDescent/itkStochasticPreconditionedGradientDescentOptimizer.h [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+ *
+ *     \f[ a(k) =  a / (A + k + 1)^alpha \f].
+ *
+ * It is very suitable to be used in combination with a stochastic estimate
+ * of the gradient \f$dC/dx\f$. For example, in image registration problems it is
+ * often advantageous to compute the metric derivative (\f$dC/dx\f$) on a new set
+ * of randomly selected image samples in each iteration. You may set the parameter
+ * \c NewSamplesEveryIteration to \c "true" to achieve this effect.
+ * For more information on this strategy, you may have a look at:
+ *
```
Comment (YuchuanQiao) [inline Components/Optimizers/PreconditionedGradientDescent/itkStochasticPreconditionedGradientDescentOptimizer.h]: Qiao Y, Lelieveldt BP, Staring M. An efficient preconditioner for stochastic gradient descent optimization of image registration. IEEE transactions on medical imaging. 2019 Feb 11.

File: Components/Optimizers/PreconditionedStochasticGradientDescent/itkPreconditionedASGDOptimizer.h [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+*
+*     \f[ t_{k+1} = [ t_k + sigmoid( -g_k^T g_{k-1} ) ]^+ \f]
+*
+* where \f$g_k\f$ equals \f$dC/dx\f$ at iteration \f$k\f$.
+* For \f$t_0\f$ the InitialTime is used, which is defined in the
+* the superclass (StandardGradientDescentOptimizer). Whereas in the
+* superclass this parameter is superfluous, in this class it makes sense.
+*
+* This method is described in the following references:
+*
```
Comment (YuchuanQiao) [inline Components/Optimizers/PreconditionedStochasticGradientDescent/itkPreconditionedASGDOptimizer.h]: Qiao Y, Lelieveldt BP, Staring M. An efficient preconditioner for stochastic gradient descent optimization of image registration. IEEE transactions on medical imaging. 2019 Feb 11.

**Comment by mstaring** [comment]
I updated some refs. I think more can be improved, many seem a copy-paste from a previous file. But that can be done after the merge in a new PR.

**Pull request merged**